sdk/storage/azure_storage_blob/tests/blob_client.rs

```diff
@@ -58,3 +98,162 @@ async fn test_get_blob_properties_invalid_container() -> Result<(), Box<dyn Erro
 
     Ok(())
 }
+
+#[recorded::test]
+async fn test_download_blob(ctx: TestContext) -> Result<(), Box<dyn Error>> {
+    // Recording Setup
+    let recording = ctx.recording();
+    let mut options = BlobClientOptions::default();
+    recording.instrument(&mut options.client_options);
+
+    // Setup
+    let endpoint = format!(
+        "https://{}.blob.core.windows.net/",
+        recording.var("AZURE_STORAGE_ACCOUNT_NAME", None).as_str()
+    );
+
+    // Act
+    let container_client = BlobContainerClient::new(
+        &endpoint,
+        String::from("testcontainer2"),
+        recording.credential(),
+        Some(options.clone()),
+    )?;
+    container_client.create_container(None).await?;
+
+    let blob_client = BlobClient::new(
+        &endpoint,
+        String::from("testcontainer2"),
+        String::from("test_blob.txt"),
+        recording.credential(),
+        Some(options),
+    )?;
+    let data = b"test download content";
+    blob_client
+        .upload_blob(
+            RequestContent::from(data.to_vec()),
+            true,
+            i64::try_from(data.len())?,
+            None,
+        )
+        .await?;
+    let response = blob_client
+        .download_blob(Some(BlobBlobClientDownloadOptions::default()))
+        .await?;
+
+    // Assert
+    let (status_code, headers, response_body) = response.deconstruct();
+    assert!(status_code.is_success());
+    assert_eq!(
+        "21",
+        headers.get_str(&HeaderName::from_static("content-length"))?
+    );
+    assert_eq!(Bytes::from_static(data), response_body.collect().await?);
+
+    container_client.delete_container(None).await?;
```

Comment:
Is there a way to ensure this runs eve if the test fails on an earlier line?

Reply:
That's a good point. I know we chatted offline about `try-finally` but in fact Rust does not have that control flow.

@heaths Heath, is there an idiomatic way to handle cleaning up test resources in both end of test and error cases? I couldn't find anything in my cursory glance through Cosmos and KeyVault tests.

One possible idea is making a struct to hold reference to all the clients necessary for cleanup, then implementing the [Drop](https://doc.rust-lang.org/rust-by-example/trait/drop.html) trait?

Reply:
There are no exceptions in Rust, but you still control the flow. Don't use the `?` operator and instead capture the result, do something, then react to the result e.g.,

```rust
let result = response_body.collect().await;
// ignore error on next line
let _ = container_client.delete_container(None).await;
assert_eq!(Bytes::from_data(data), result?); // notice the questio mark here, or use `.unwrap()` and let the test panic / fail
```

It's more like Go or C/C++. Handle the errors when you want to.

Reply:
I didn't see a change for this one.

Reply:
Hi Heath, we are still having discussions internally with how we want to test resources, but will definitely be thinking on this in the next PR.